File: Largest-K-Divisible-Subsequence.py
class Solution:
  def solve(self, array, modulo):
    length = len(array)

    dp = [[-1 for _ in range(modulo)] for _ in range(length + 1)]
    dp[0][0] = 0

    for index in range(length):
      dp[index + 1] = dp[index][:]

      for remainder in range(modulo):
        if dp[index][remainder] != -1:
          dp[index + 1][(remainder + array[index]) % modulo] = \
            max(dp[index + 1][(remainder + array[index]) % modulo],
                dp[index][remainder] + array[index])

    return dp[length][0]

File: Largest-K-Divisible-Subsequence.cpp
const int kImpossible = -1;

int solve(std::vector<int>& array, int modulo) {
  const int length = array.size();

  // dp[prefix length][remainder] = maximum sum
  //   of a subsequence of a given prefix with given modulo remainder
  std::vector<std::vector<int64_t>> dp(length + 1,
    std::vector<int64_t>(modulo, kImpossible));
  dp[0][0] = 0;

  for (int index = 0; index < length; ++index) {
    // if we do NOT take array[index] into subsequence sum
    std::copy(dp[index].begin(), dp[index].end(), dp[index + 1].begin());

    // if we DO take array[index] into subsequence sum
    for (int remainder = 0; remainder < modulo; ++remainder) {
      if (dp[index][remainder] != kImpossible) {
        dp[index + 1][(remainder + array[index]) % modulo] =
          std::max(dp[index + 1][(remainder + array[index]) % modulo],
                   dp[index][remainder] + array[index]);
      }
    }
  }

  // maximum sum of a subsequence of an entire array divisible by modulo
  return dp[length][0];
}
